### ISSUE TITLE
Unbinding $this of closure is deprecated

### ISSUE THREAD

**Issue by kylekatarnls** [laravel/framework#29411]
- Laravel Version: 5.8 / 6.0
- PHP Version: 7.4.0 / 8.0.0

### Description:

PHP deprecate closure unbinding, so the following line:
https://github.com/laravel/framework/blob/master/src/Illuminate/Support/Traits/Macroable.php#L110

throws:
`Unbinding $this of closure is deprecated`

since PHP 7.4.0

As having a similar macro system in Carbon, I'm working on possible alternatives.

Do you plan to support PHP 7.4 and 8? And to still support mixin/macro in those new versions?

**Comment by derekmd** [comment]
Relevant links:

* https://github.com/php/php-src/commit/28cf080701082cbb1848d511d38cc1ecd060a752#diff-833c0daf9450be2a512d6cecf4351b6c
* https://wiki.php.net/rfc/deprecations_php_7_4#unbinding_this_from_non-static_closures
* https://news-web.php.net/php.internals/106158 (bottom paragraph)

Laravel `Macroable::macro()` calls in user-land will likely need to be refactored into class-based mixins. `Macroable@__call()` would needs its `if ($macro instanceof Closure) {` conditional deprecated.

**Comment by driesvints** [comment]
Hmm. We should try to get this fixed before v6.0 is out since PHP 7.4 drops in fall this year.

**Comment by kylekatarnls** [comment]
@driesvints The option I took for next Carbon release is muting the errors with `@`. Of course it just postpones the problem to a later version. But it would give a few more time.

**Comment by driesvints** [comment]
@kylekatarnls yeah I'd opt to actually provide a fix already if we can.

**Comment by kylekatarnls** [comment]
Here is for the quick fix: #29418

But some testing utils throws other deprecation notices with PHP 7.4 in Travis-CI. I will try to fix them.

No fix for Laravel 5.8?

**Comment by xmoroccan** [comment]
Using error control symbol `@` is a bad practice, it can prevent other errors from being shown and/or handled, and you'll have a more time just to know what the heck is happening.
From PHP docs:

> Warning: Currently the "@" error-control operator prefix will even disable error reporting for critical errors that will terminate script execution. Among other things, this means that if you use "@" to suppress errors from a certain function and either it isn't available or has been mistyped, the script will die right there with no indication as to why.

IMO, and from PR #29418 you shouldn't use `@` before a `call_user_func_array`, because we'll never know if the closure or the function typed by the developer is working and does not have any typos or any kind of errors!

**Comment by driesvints** [comment]
@kylekatarnls I've closed your temp fix pr. Let's try to fix the actual problem before we release 6.0

**Comment by driesvints** [comment]
> No fix for Laravel 5.8?

No, we can't make breaking changes here.

**Comment by driesvints** [comment]
Okay so here's the gist: everything will still work as is except for static macro calls. When you perform a static macro call which is a closure you'll need to make sure it's also marked the closure as static. Only the unbinding of $this for **non-static** closures is deprecated. The error message above is actually a bit misleading. 

So basically the following won't work anymore:

```php
FooClass::macro('tryStatic', function () {
    return 'foo';
});

FooClass::tryStatic();
```

Because the closure is non-static you can't unbind `$this` anymore. But if you do the following, everything will work as expected:


```php
FooClass::macro('tryStatic', static function () {
    return 'foo';
});

FooClass::tryStatic();
```

Notice the static keyword. The closure is marked as static now and `$this` cannot be called from within anymore. 

I've sent in a PR here: https://github.com/laravel/framework/pull/29482

**Comment by faliure** [comment]
In Macroable.php:110 we're inside __call, I don't see how we'd get that error there: `return call_user_func_array($macro->bindTo($this, static::class), $parameters);`. It's binding to $this, in a non-static context, $this cannot be null.

Maybe the error comes from line 84? That'd make more sense, unless I'm missing something big.

**Comment by GrahamCampbell** [comment]
This will be fixed in PHP 7.4.0 RC1.
